controller/controller.go

```diff
@@ -1287,6 +1323,19 @@ func (ctrl *ProvisionController) provisionClaimOperation(claim *v1.PersistentVol
 		}
 		err = fmt.Errorf("failed to provision volume with StorageClass %q: %v", claimClass, err)
 		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, "ProvisioningFailed", err.Error())
+		if _, ok := claim.Annotations[annSelectedNode]; ok && result == ProvisioningFinished {
+			// For dynamic PV provisioning with delay binding, provisioner may fail
+			// because node is wrong (permanent error) or currently unusable (not enough
+			// capacity). In the later case, the provisioner must decide whether it wants
+			// to keep retrying (i.e. return ProvisioningNoChange) or give up on the
+			// node (i.e. return ProvisioningFinished). This code here only gives
+			// up on the node when it gets ProvisioningFinished as result.
+			//
+			// `selectedNode` must be removed to notify scheduler to schedule again.
+			// No need to handle error here, because if `selectedNode` is failed to
+			// be removed, next retry will remove it again.
+			ctrl.rescheduleProvisioning(claim)
```

Comment:
It's slightly more complicated. If the annotation is not removed, the provisioner enqueues the PVC back to the worker queue, as it is done below with `return result, err`. This part is OK.

If the annotation **is** removed, the provisioner basically waits for scheduler to re-schedule the PVC/Pod. But with `return result, err` below, the PVC is returned back to the provisioner workqueue with exp. backoff. This looks bad, however, the next `syncClaim()` removes the PVC from queue, as the PVC does not pass `shouldProvision` check. Which is correct, in the end.

The behavior is probably better than returning a new "error, but don't re-queue" value, but the comment should be updated.

Reply:
This comment and code is straight from @cofyc's original PR, so I have to guess a bit what was meant: my interpretation is slightly different, I think "no need to handle error here" refers to an error in `ctrl.rescheduleProvisioning`. What the comment is saying is that such a problem can be ignored because the next scheduling attempt will try again and eventually the annotation should get removed.

> But with return result, err below, the PVC is returned back to the provisioner workqueue with exp. backoff.

Is that really the case? `result` is `ProvisioningFinished`, doesn't that mean that the error is final and thus no future retries should be attempted?

Reply:
`ProvisioningFinished` means that provisioning is retried as long as user wants the volume (i.e. PVC exists). The provisioner stops trying when PVC is deleted.

`ProvisioningInBackground` means the provisioner re-tries even after PVC is deleted, to catch the final error and make sure the volume that was being provisioned is not leaked.

Reply:
> It's slightly more complicated. If the annotation is not removed, the provisioner enqueues the PVC back to the worker queue, as it is done below with return result, err. 

What happens if we return with `ProvisioningFinished, nil` as in https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1219-L1221?

Is that then "cleanly" removing the volume from the queue? This might be better than falling through to the `return result, err` if unsetting the label succeeded.

Reply:
> Is that then "cleanly" removing the volume from the queue?

Probably need to address this TODO for that: https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/0a77d7821695135d58fa7c59af6c042ae24a6adc/controller/controller.go#L959

Reply:
> What happens if we return with ProvisioningFinished, nil as in

It will mess up provisioner metrics - the call will be recorded as successful.

Reply:
Looking at the code, should we pass err down to `syncClaim()`, record failure in metrics there and only if the status is `ProvisioningFinished` then return nil from `syncClaim()`? It will not requeue the claim.

Reply:
> It will mess up provisioner metrics - the call will be recorded as successful.

There are already other places where `provisionClaimOperation` returns `ProvisioningFinished, nil` (https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1219-L1221, https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1247-L1251).

Do those also mess up metrics?

Another one does it right after emitting an error event: https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1234-L1236

> Looking at the code, should we pass err down to syncClaim(), record failure in metrics there and only if the status is ProvisioningFinished then return nil from syncClaim()? 

I think that makes sense. Let me recap:
- never return `ProvisioningFinished, nil` from `provisionClaimOperation` unless provisioning really succeeded, which implies updating some of the existing locations such that they return an error
- in `syncClaim`, insert `if status == ProvisioningFinished { return nil }` before https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1029

Hmm, but doesn't that then break the exponential backoff for final errors (https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L895-L910)?

Reply:
> Do those also mess up metrics?

The first one does not - volume already exists, so provisioning *was* successful
The other ones are indeed wrong and mess up metrics.

> * never return ProvisioningFinished, nil from provisionClaimOperation unless provisioning really succeeded, which implies updating some of the existing locations such that they return an error

Ack

> * in syncClaim, insert `if status == ProvisioningFinished { return nil }` before
https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/7c59ae8f420d068d22240b90075c8699d64c7a21/controller/controller.go#L1029

> Hmm, but doesn't that then break the exponential backoff for final errors (

Ack, it breaks exp. backoff.

What about `if status == ProvisioningReschedule { return nil }` at the same place?

Reply:
> The first one does not - volume already exists, so provisioning was successful

But isn't this success then going to be counted again? Do we have tests and documentation for these metrices somewhere?

> What about `if status == ProvisioningReschedule { return nil }` at the same place?

Not sure. I suspect that it might be easier to to use a special error instead of `nil` in `provisionClaimOperation` to distinguish "volume provisioned" from "volume not provisioned" and then check for that in the callers in those places where it matters.

Reply:
> Do those also mess up metrics?

Of these special cases, only one was covered by the unit testing ("provisioned ignored volume"), but even that one case didn't check the statistics. Those indeed reported "two volumes provisioned successfully" although one of the two was ignored instead of provisioned.

> I suspect that it might be easier to to use a special error instead of nil in provisionClaimOperation to distinguish "volume provisioned" from "volume not provisioned" and then check for that in the callers in those places where it matters.

I have added checking of metrics values and then (in a separate commit) fixed those special cases by implementing this approach. I also found one code path where a failed provisioning wasn't counted.

@jsafrane : I've structured the changes so that the initial set of commits are separate fixes for the existing code before adding the rescheduling in the final commit. The best approach for reviewing this PR is commit-by-commit.